File: 8/a.ts
// const input = await Bun.file("test.txt").text();
const input = await Bun.file("input.txt").text();

let res = 0;

const lines = input.split("\n");

const sequence = lines.shift()!;
lines.shift();

const obj: Record<
  string,
  {
    R: string;
    L: string;
  }
> = {};

lines.forEach((line) => {
  const [key, value] = line.split(" = ");
  obj[key] = {
    L: value.slice(1, 4),
    R: value.slice(6, 9),
  };
});

let i = 0;
let current = "AAA";
while (true) {
  if (i > sequence.length - 1) i = 0;
  let a = sequence[i];
  current = obj[current][a as "R" | "L"];
  res++;
  i++;
  console.log(current, a, i);
  if (current === "ZZZ") break;
  // await new Promise((r) => setTimeout(r, 1000));
}

console.log(res);

File: 8/a.go
package main

import (
	"fmt"
	"os"
	"strings"
)

type Node struct {
	right string
	left  string
}

func main() {
	file, _ := os.ReadFile("input.txt")

	lines := strings.Split(string(file), "\n")

	sequence := lines[0]
	nodesMap := make(map[string]Node)
	for i := 2; i < len(lines); i++ {
		key, rawValue, _ := strings.Cut(lines[i], " = ")
		nodesMap[key] = Node{left: rawValue[1:4], right: rawValue[6:9]}
	}

	i := 0
	current := "AAA"

	for {
		newCurrent := nodesMap[current].left
		if sequence[i%len(sequence)] == 'R' {
			newCurrent = nodesMap[current].right
		}
		current = newCurrent
		i++
		if current == "ZZZ" {
			break
		}
	}

	fmt.Println(i)
}
